### PR TITLE
Provide a customResources alternative that only accepts resourceType as an argument

### PR DESCRIPTION
**Pull request by iocanel**
Resolves: #2605

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Review by manusa — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 1]
```diff
@@ -99,6 +99,49 @@
...
+  /**
+   * Typed API for managing CustomResources. You would need to provide POJOs for
+   * CustomResource into this and with it you would be able to instantiate a client
+   * specific to CustomResource.
+   *
+   * <p>
+   *   Note: your CustomResource POJO (T in this context) must implement
+   *   <a href="https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Namespaced.java">
+   *     io.fabric8.kubernetes.api.model.Namespaced
+   *   </a> if it is a Namespaced scoped resource.
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:111]:
```suggestion
   *   {@link io.fabric8.kubernetes.api.model.Namespaced} if it is a namespace-scoped resource.
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 2]
```diff
@@ -99,6 +99,49 @@
...
+   * </p>
+   *
+   * @param resourceType Class for CustomResource
+   * @param listClass Class for list object for CustomResource
+   * @param <T> T type represents CustomResource type. If it's Namespaced resource, it must implement
+   *           io.fabric8.kubernetes.api.model.Namespaced
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:117]:
```suggestion
   *           {@link io.fabric8.kubernetes.api.model.Namespaced}
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 3]
```diff
@@ -99,6 +99,49 @@
...
+   * <p>
+   *   Note: your CustomResource POJO (T in this context) must implement
+   *   <a href="https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Namespaced.java">
+   *     io.fabric8.kubernetes.api.model.Namespaced
+   *   </a> if it is a Namespaced scoped resource.
+   * </p>
+   *
+   * @param resourceType Class for CustomResource
+   * @param listClass Class for list object for CustomResource
+   * @param <T> T type represents CustomResource type. If it's Namespaced resource, it must implement
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:116]:
```suggestion
   * @param <T> T type represents CustomResource type. If it's a namespaced resource, it must implement
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 4]
```diff
@@ -99,6 +99,49 @@
...
+  /**
+   * Typed API for managing CustomResources. You would need to provide POJOs for
+   * CustomResource into this and with it you would be able to instantiate a client
+   * specific to CustomResource.
+   *
+   * <p>
+   *   Note: your CustomResource POJO (T in this context) must implement
+   *   <a href="https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/Namespaced.java">
+   *     io.fabric8.kubernetes.api.model.Namespaced
+   *   </a> if it is a Namespaced scoped resource.
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:132]:
```suggestion
   *   {@link io.fabric8.kubernetes.api.model.Namespaced} if it is a namespace-scoped resource.
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 5]
```diff
@@ -99,6 +99,49 @@
...
+   * </p>
+   *
+   * @param resourceType Class for CustomResource
+   * @param listClass Class for list object for CustomResource
+   * @param <T> T type represents CustomResource type. If it's Namespaced resource, it must implement
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:137]:
```suggestion
   * @param <T> T type represents CustomResource type. If it's a namespace-scoped resource, it must implement
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 6]
```diff
@@ -99,6 +99,49 @@
...
+   *           io.fabric8.kubernetes.api.model.Namespaced
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:138]:
```suggestion
   *           {@link io.fabric8.kubernetes.api.model.Namespaced}
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java [hunk 7]
```diff
@@ -15,7 +15,15 @@
...
 
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesResource;
+import io.fabric8.kubernetes.api.model.KubernetesResourceList;
+import io.fabric8.kubernetes.api.model.Namespaced;
 import io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinition;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
+import io.fabric8.kubernetes.client.utils.Pluralize;
+import io.fabric8.zjsonpatch.internal.guava.Strings;
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java:26]: Is that a good idea to import this class? I've thought about using it as well but it looks really weird… 😄

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Pluralize.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (C) ${project.inceptionYear} ${owner}
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/Pluralize.java:2]: 😄

**Review by metacosm — CHANGES_REQUESTED** [review]

**Comment by rohanKanojia** [comment]
Pluralize added in this PR seems to be taken from Sundrio:

https://github.com/sundrio/sundrio/blob/master/codegen/src/main/java/io/sundr/codegen/functions/Pluralize.java 

Is it possible to reuse this from sundrio?

**Comment by metacosm** [comment]
> Is it possible to reuse this from sundrio?

Then again, it might be better not to tie the runtime to sundrio?

**Comment by rohanKanojia** [comment]
ohk, let's keep it here then.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java (hunk 7 above)
Comment (iocanel) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/CustomResourceDefinitionContext.java:26]: Noi, its not and in fact it was causing issues, with the OSGi tests.

**Comment by iocanel** [comment]
> Pluralize added in this PR seems to be taken from Sundrio:
> 
> https://github.com/sundrio/sundrio/blob/master/codegen/src/main/java/io/sundr/codegen/functions/Pluralize.java
> 
> Is it possible to reuse this from sundrio?

I'd rather not have a runtime dependency in sundrio.

**Review by metacosm — APPROVED** [review]
LGTM

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java [hunk 9]
```diff
@@ -99,6 +99,45 @@
...
+   * CustomResource into this and with it you would be able to instantiate a client
+   * specific to CustomResource.
+   *
+   * <p>
+   *   Note: your CustomResource POJO (T in this context) must implement
+   *   {@link io.fabric8.kubernetes.api.model.Namespaced} if it is a namespace-scoped resource.
+   * </p>
+   *
+   * @param resourceType Class for CustomResource
+   * @param listClass Class for list object for CustomResource
```
Comment (metacosm) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java:113]:
```suggestion
```

**Review by manusa — APPROVED** [review]

**Comment by manusa** [comment]
[merge]

**Pull request merged**